src/Query/AggregationBuilder.php

```diff
@@ -0,0 +1,98 @@
+<?php
+
+declare(strict_types=1);
+
+namespace MongoDB\Laravel\Query;
+
+use Illuminate\Support\Collection as LaravelCollection;
+use Illuminate\Support\LazyCollection;
+use InvalidArgumentException;
+use Iterator;
+use MongoDB\Builder\BuilderEncoder;
+use MongoDB\Builder\Stage\FluentFactoryTrait;
+use MongoDB\Collection as MongoDBCollection;
+use MongoDB\Driver\CursorInterface;
+use MongoDB\Laravel\Collection as LaravelMongoDBCollection;
+
+use function array_replace;
+use function collect;
+use function sprintf;
+use function str_starts_with;
+
+class AggregationBuilder
+{
+    use FluentFactoryTrait;
+
+    public function __construct(
+        private MongoDBCollection|LaravelMongoDBCollection $collection,
+        private readonly array $options = [],
+    ) {
+    }
+
+    /**
+     * Add a stage without using the builder. Necessary if the stage is built
+     * outside the builder, or it is not yet supported by the library.
+     */
+    public function addRawStage(string $operator, mixed $value): static
+    {
+        if (! str_starts_with($operator, '$')) {
+            throw new InvalidArgumentException(sprintf('The stage name "%s" is invalid. It must start with a "$" sign.', $operator));
+        }
+
+        $this->pipeline[] = [$operator => $value];
```

Comment:
`FluentFactoryTrait::$pipeline` is annotated as a `list<StageInterface>`. Does this violate that contract?

See: https://github.com/mongodb/mongo-php-builder/blob/aa4f34cde35608b328f23d8fa6270b7cafbe5b18/src/Builder/Stage/FluentFactoryTrait.php#L36C14-L36C34

More generally, I would have expected this kind of escape hatch to exist in the builder itself if at all.

Reply:
Exact: https://github.com/mongodb/mongo-php-builder/pull/71

Reply:
Do you think I should add a constructor to `FluentFactoryTrait`? I'm not sure.

Reply:
What do you mean by "add a constructor"? AFAIK, traits cannot have constructors and I don't see why you'd need an initialization method.

Reply:
> What do you mean by "add a constructor"? AFAIK, traits cannot have constructors and I don't see why you'd need an initialization method.

Indeed, the constructor initialization is not necessary.


> Are you suggesting putting `addRawStage()` in the trait?

No, keeping it for Laravel only for now.